cbmc/proofs/IotMqtt_SubscribeSync/IotMqtt_SubscribeSync_harness.c

```diff
@@ -0,0 +1,157 @@
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file IotMqtt_SubscribeSync_harness.c
+ * @brief Implements the proof harness for IotMqtt_SubscribeSync function.
+ */
+
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+
+#include "mqtt_state.h"
+
+typedef bool ( * MatchFunction_t )( const IotLink_t * const pOperationLink,
+                                    void * pCompare );
+
+typedef void ( * FreeElementFunction_t )( void * pData );
+
+static unsigned int flagSemaphore;
+
+/**
+ * We constrain the return values of this function because it
+ * is checked by assertions in the MQTT code.
+ */
+IotTaskPoolError_t IotTaskPool_TryCancel( IotTaskPool_t taskPool,
+                                          IotTaskPoolJob_t job,
+                                          IotTaskPoolJobStatus_t * const pStatus )
+{
+    if( ( taskPool == NULL ) || ( job == NULL ) )
+    {
+        return IOT_TASKPOOL_BAD_PARAMETER;
+    }
+
+    return IOT_TASKPOOL_SUCCESS;
+}
+
+/**
+ * _IotMqtt_NextPacketIdentifier calls Atomic_Add_u32, which receives
+ * a volatile variable as input. Thus, CBMC will always consider that
+ * Atomic_Add_u32 will operate over nondetermistic values and raise
+ * an unsigned integer overflow failure. However, developers have reported
+ * that the use of this overflow is part of the function implementation.
+ * In order to mirror _IotMqtt_NextPacketIdentifier behaviour and avoid
+ * spurious alarms, we stub out this function to always
+ * return a nondetermistic odd value.
+ */
+uint16_t _IotMqtt_NextPacketIdentifier( void )
+{
+    uint16_t id;
+
+    /* Packet identifiers will follow the sequence 1,3,5...65535,1,3,5... */
+    __CPROVER_assume( id & 0x0001 == 1 );
+
+    return id;
+}
+
+/**
+ * We abstract these functions for performance reasons. In their original
+ * implementation, CBMC ended up creating byte extracts for all possible
+ * objects, due to the polymorphic nature of the linked list. We assume
+ * that these functions are memory safe, and we free and havoc the list to
+ * demonstrate that no subsequent code makes use of the values in the list.
+ */
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+                                     MatchFunction_t isMatch,
+                                     void * pMatch,
+                                     FreeElementFunction_t freeElement,
+                                     size_t linkOffset )
+{
```

Comment:
These signatures should match exactly what is in the source code.  Is this refactoring of MatchFunction_t and FreeElementFunction_t your idea or was it done in the source code?  Should we change this in all proofs?